src/probnum/diffeq/perturbedsolvers/perturbedstepsolution.py

```diff
@@ -0,0 +1,74 @@
+import numpy as np
+from scipy.integrate._ivp import rk
+
+from probnum import _randomvariablelist, diffeq, randvars
+from probnum.filtsmooth.timeseriesposterior import (
+    DenseOutputLocationArgType,
+    DenseOutputValueType,
+)
+
+
+class PerturbedStepSolution(diffeq.ODESolution):
+    """Output of NoisyStepSolver."""
+
+    def __init__(
+        self,
+        scales: np.array,
+        times: np.array,
+        rvs: np.array,
+        interpolants: rk.RkDenseOutput,
+    ):
+        self.scales = scales
+        self.interpolants = interpolants
+        super().__init__(times, rvs)
+
+    def __call__(self, t: DenseOutputLocationArgType) -> DenseOutputValueType:
+        if not np.isscalar(t):
+
+            # recursive evaluation (t can now be any array, not just length 1!)
+            return _randomvariablelist._RandomVariableList(
+                [self.__call__(t_pt) for t_pt in np.asarray(t)]
+            )
+
+        # Access last state directly.
+        if t == self.locations[-1]:
+            res = self.states[-1]
+        else:
+
+            # get index of closest left interpolant (í.e. correct interpolant).
+            closest_left_t = get_interpolant(self.locations, t)
+            interpolant = self.interpolants[closest_left_t]
+            if self.locations[closest_left_t] == t:
+                res = randvars.Constant(interpolant(t))
+            else:
+                relative_time = (t - self.locations[closest_left_t]) * self.scales[
+                    closest_left_t
+                ]
+                res = randvars.Constant(
+                    interpolant(self.locations[closest_left_t] + relative_time)
+                )
+        return res
```

Comment:
I suppose your life would be easier if you implemented interpolate() instead of call() :) This has changed in ProbNum since you started, and might be worth looking at :)

Reply:
in particular, all the searching and sorting is handled and there should be no need for `get_interpolant` below :)